homeassistant/components/mqtt/notify.py

```diff
@@ -0,0 +1,393 @@
+    for key, service in services.items():  # pylint: disable=unused-variable
+        if service.device_id == device_id:
+            return True
+    return False
+
+
+async def async_get_service(
+    hass: HomeAssistant,
+    config: ConfigType,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> MqttNotificationService | None:
+    """Prepare the MQTT notification service through configuration.yaml."""
+    name = config.get(CONF_NAME)
+    if CONF_NAME not in config:
+        config[CONF_NAME] = DOMAIN
+    service_name = slugify(name or DOMAIN)
+    has_services = hass.services.has_service(notify.DOMAIN, service_name)
+    services = hass.data.setdefault(MQTT_NOTIFY_SERVICES_SETUP, {})
+    if service_name in services.keys() or has_services:
+        _LOGGER.error(
+            "Notify service '%s' is not unique, cannot register service(s)",
+            service_name,
+        )
+        return None
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    services[service_name] = MqttNotificationService(
+        hass,
+        config[CONF_COMMAND_TOPIC],
+        MqttCommandTemplate(config.get(CONF_COMMAND_TEMPLATE), hass=hass),
+        config[CONF_ENCODING],
+        name,
+        config[CONF_QOS],
+        config[CONF_RETAIN],
+        config[CONF_TARGETS],
+        config[CONF_TITLE],
+    )
+    return services[service_name]
+
+
+class MqttNotificationServiceUpdater:
+    """Add support for auto discovery updates."""
+
+    def __init__(self, hass: HomeAssistant, service: MqttNotificationService) -> None:
+        """Initialize the update service."""
+
+        async def async_discovery_update(
+            discovery_payload: DiscoveryInfoType | None,
+        ) -> None:
+            """Handle discovery update."""
+            async_dispatcher_send(
+                hass, MQTT_DISCOVERY_DONE.format(service.discovery_hash), None
+            )
+            if not discovery_payload:
+                # unregister notify service through auto discovery
+                await async_tear_down_service()
+                return
+
+            # update notify service through auto discovery
+            await service.async_update_service(discovery_payload)
+            _LOGGER.debug(
+                "Notify service %s updated has been processed",
+                service.discovery_hash,
+            )
+
+        async def async_device_removed(event):
+            """Handle the removal of a device."""
+            device_id = event.data["device_id"]
+            if (
+                event.data["action"] != "remove"
+                or device_id != service.device_id
+                or self._device_removed
+            ):
+                return
+            self._device_removed = True
+            await async_tear_down_service()
+
+        async def async_tear_down_service():
+            """Handle the removal of the service."""
+            services = hass.data.setdefault(MQTT_NOTIFY_SERVICES_SETUP, {})
```

Comment:
Instead of doing this everywhere, add a call to `notify.initialize(hass)` in `mqtt/__init__.py::async_setup` and have `notify.initialize(hass)` setup mandatory things.

Reply:
Things were centralized. There is still 2 places were initialization can happen, but the code is in one place.